table/tables/partition_test.go

```diff
@@ -714,6 +2170,87 @@ func TestIssue31721(t *testing.T) {
+
+	tk.MustExec("CREATE TABLE tkey16 (\n" +
+		"col1 INT NOT NULL," +
+		"col2 INT NOT NULL," +
+		"col3 INT NOT NULL," +
+		"col4 INT NOT NULL," +
+		"primary KEY (col1,col3)\n" +
+		")" +
+		"PARTITION BY HASH(col3) PARTITIONS 4")
+	tk.MustExec("INSERT INTO tkey16 values(1,1,1,1),(1,1,2,2),(3,3,3,3),(3,3,4,3),(4,4,4,4),(5,5,5,5),(6,6,6,6),(7,7,7,7),(8,8,8,8),(9,9,9,9),(10,10,10,5),(11,11,11,6),(12,12,12,12),(13,13,13,13),(14,14,14,14)")
+
+	err := tk.ExecToErr("ALTER TABLE tkey15 PARTITION BY KEY(col3) PARTITIONS 4")
+	require.Regexp(t, "alter table partition is unsupported", err)
+	err = tk.ExecToErr("ALTER TABLE tkey14 ADD PARTITION PARTITIONS 1")
+	require.Regexp(t, "Unsupported add partitions", err)
+	err = tk.ExecToErr("ALTER TABLE tkey14 DROP PARTITION p4")
+	require.Regexp(t, "DROP PARTITION can only be used on RANGE/LIST partitions", err)
+	tk.MustExec("ALTER TABLE tkey14 TRUNCATE PARTITION p3")
+	tk.MustQuery("SELECT COUNT(*) FROM tkey14 partition(p3)").Check(testkit.Rows("0"))
+	err = tk.ExecToErr("ALTER TABLE tkey16 COALESCE PARTITION 2")
+	require.Regexp(t, "Unsupported coalesce partitions", err)
+	tk.MustExec("ALTER TABLE tkey14 ANALYZE PARTITION p3")
+	err = tk.ExecToErr("ALTER TABLE tkey14 CHECK PARTITION p2")
+	require.Regexp(t, "Unsupported check partition", err)
+	err = tk.ExecToErr("ALTER TABLE tkey14 OPTIMIZE PARTITION p2")
+	require.Regexp(t, "Unsupported optimize partition", err)
+	err = tk.ExecToErr("ALTER TABLE tkey14 REBUILD PARTITION p2")
+	require.Regexp(t, "Unsupported rebuild partition", err)
+	err = tk.ExecToErr("ALTER TABLE tkey14 EXCHANGE PARTITION p3 WITH TABLE tkey15")
+	require.Regexp(t, "Found a row that does not match the partition", err)
+
+	tk.MustExec("DELETE FROM tkey15")
+	tk.MustExec("INSERT INTO tkey15 VALUES(11,11,11,6)")
+	tk.MustExec("ALTER TABLE tkey14 EXCHANGE PARTITION p3 WITH TABLE tkey15")
+	tk.MustQuery("SELECT * FROM tkey14 partition(p3)").Check(testkit.Rows("11 11 11 6"))
+
+	err = tk.ExecToErr("ALTER TABLE tkey16 REORGANIZE PARTITION")
+	require.Regexp(t, "Unsupported reorganize partition", err)
+	err = tk.ExecToErr("ALTER TABLE tkey16 REORGANIZE PARTITION p0 INTO (PARTITION p0,PARTITION p1)")
+	require.Regexp(t, "Unsupported reorganize partition", err)
+	err = tk.ExecToErr("ALTER TABLE tkey16 REORGANIZE PARTITION p0 INTO (PARTITION p0)")
+	require.Regexp(t, "Unsupported reorganize partition", err)
+	err = tk.ExecToErr("ALTER TABLE tkey16 REORGANIZE PARTITION p0 INTO (PARTITION p4)")
+	require.Regexp(t, "Unsupported reorganize partition", err)
+	err = tk.ExecToErr("ALTER TABLE tkey16 REMOVE PARTITIONING")
+	require.Regexp(t, "Unsupported remove partitioning", err)
+
+	tk.MustExec("CREATE TABLE tkey17 (" +
+		"id INT NOT NULL PRIMARY KEY," +
+		"name VARCHAR(20)" +
+		")" +
+		"PARTITION BY KEY()" +
+		"PARTITIONS 2")
+	result := tk.MustQuery("show warnings")
+	result.CheckContains("Unsupported partition type KEY, treat as normal table")
```

Comment:
This would work in MySQL:
```
mysql> CREATE TABLE tkey17 (id INT NOT NULL PRIMARY KEY,name VARCHAR(20)) PARTITION BY KEY() PARTITIONS 2;
Query OK, 0 rows affected (0,02 sec)
```

Would it be easy to support it in this PR as well?

Reply:
Listed as an restriction in https://github.com/pingcap/tidb/issues/41364

Reply:
I have listed it .